### PR TITLE
[ios, macos] Fix visible coordinates when boundaries cross international datelline

### PR DESCRIPTION
**Pull request by fabian-guerra**
Adds documentation to clarify how to use coordinate bounds that cross the anti-meridian.

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -2667,8 +2667,21 @@ - (void)_setVisibleCoordinates:(const CLLocationCoordinate2D *)coordinates count
...
+    BOOL boundariesCrossDateTime = NO;
+    if (count == 4 && coordinates[1].longitude > 0 && coordinates[3].longitude < 0) {
+        boundariesCrossDateTime = YES;
+        latLngs = [self normalizeCoordinatesForDateTimeChangeSW:coordinates[1] ne:coordinates[3]];
+        
+    }
 
     mbgl::CameraOptions cameraOptions = _mbglMap->cameraForLatLngs(latLngs, padding);
+    
+    if (boundariesCrossDateTime) {
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This code base calls it the “[antimeridian](https://en.wikipedia.org/wiki/Antimeridian)”, not the “date time”.

File: platform/darwin/src/MGLGeometry_Private.h [hunk 2]
```diff
@@ -123,3 +123,48 @@ NS_INLINE MGLRadianCoordinate2D MGLRadianCoordinateAtDistanceFacingDirection(MGL
...
+
+/** Returns the coordinate at a given fraction between two coordinates.
+ 
+ @param origin Origin `CLLocationCoordinate2D`.
+ @param destination Destination `CLLocationCoordinate2D`.
+ @param fraction Fraction between coordinates (0 = origin, 0.5 = middle, 1 = destination).
+ 
+ @return A coordinate at given fraction. */
+// Ported from http://www.movable-type.co.uk/scripts/latlong.html
+NS_INLINE CLLocationCoordinate2D MGLCLCoordinateAtFraction(CLLocationCoordinate2D origin, CLLocationCoordinate2D destination, double fraction) {
```
Comment (1ec5) [inline platform/darwin/src/MGLGeometry_Private.h]: This is a lot of code to inline. Move the body of the function to MGLGeometry.mm.

**Comment by 1ec5** [comment]
> Can you investigate how that works and adapt Transform to do likewise?

Actually, it looks like the method affected by this PR uses `-cameraForCameraOptions:`, so I’d look into whether `mbgl::Map::cameraForLatLngs()` needs to know how to straddle the antimeridian.

**Comment by friedbunny** [comment]
This seems like good candidate for testing; perhaps in `MGLGeometryTests`.

**Comment by fabian-guerra** [comment]
@1ec5 I changed the implementation to follow this https://github.com/mapbox/mapbox-gl-native/blob/release-ios-v3.6.0-android-v5.1.0/platform/darwin/src/MGLPolyline.h#L36 that led me to avoid adding new functions.

File: platform/ios/src/MGLMapView.mm [hunk 3]
```diff
@@ -2667,6 +2667,11 @@ - (void)_setVisibleCoordinates:(const CLLocationCoordinate2D *)coordinates count
     {
         latLngs.push_back({coordinates[i].latitude, coordinates[i].longitude});
     }
+    
+    if (count == 4 && coordinates[1].longitude > 0 && coordinates[3].longitude < 0) {
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This public method is intended to work for an arbitrary number of coordinates.

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -2707,6 +2712,19 @@ - (void)_setVisibleCoordinates:(const CLLocationCoordinate2D *)coordinates count
     [self didChangeValueForKey:@"visibleCoordinateBounds"];
 }
 
+- (std::vector<mbgl::LatLng>)normalizeCoordinatesForAntimeridianSW:(CLLocationCoordinate2D)sw ne:(CLLocationCoordinate2D)ne
+{
+    CLLocationDegrees longitude = (180 + (180 - fabs(sw.longitude))) * -1;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: If I’m not mistaken, the `mbgl::LatLngBounds` class can handle spanning the antimeridian – that’s how the existing shape annotation functionality is able to handle the antimeridian gracefully. See the `mbgl::LatLngBounds::extend()` method in particular.

**Comment by fabian-guerra** [comment]
@1ec5 Giving a second thought I'm thinking the best solution for this is keep the consistency across our API by adding the proper documentation to `setVisibleCoordinateBounds` just like https://github.com/mapbox/mapbox-gl-native/blob/release-ios-v3.6.0-android-v5.1.0/platform/darwin/src/MGLPolyline.h#L36 in Polyline we are not responsible for extending the coordinates. Thoughts?

/cc @boundsj

**Comment by 1ec5** [comment]
I’m not sure I follow. That documentation states that MGLPolyline correctly handles straddling the antimeridian on the map as long as you specify some coordinates beyond ±180° longitude. #9731 is about `-[MGLMapView setVisibleCoordinateBounds:]` fails to straddle the antimeridian and instead shows “the other side” of the world.

**Comment by fabian-guerra** [comment]
What I mean is if devs also set `setVisibleCoordinateBounds` beyond ±180º it will work.

**Comment by 1ec5** [comment]
Oh, I see what you mean. If setting a coordinate bounds that goes beyond ±180° already works, then that’s great. The remaining issue is whether the developer should get the same behavior by setting a “backwards” coordinate bounds in which `sw.longitude > ne.longitude`. As long as the ±180° approach works, it isn’t as important to address the remaining issue, at least not while #4664 is still open.

**Comment by fabian-guerra** [comment]
I added the documentation comments to clarify how to use `bounds` that cross the anti-meridian. We already support extending longitudes beyond ±180°.

File: platform/ios/src/MGLMapView.h [hunk 5]
```diff
@@ -668,12 +668,24 @@ MGL_EXPORT IB_DESIGNABLE
  Changing the value of this property updates the receiver immediately. If you
  want to animate the change, call `-setVisibleCoordinateBounds:animated:`
  instead.
+ 
+ To make the visible bounds go across the antimeridian or international date line,
+ specify some longitudes less than −180 degrees or greater than 180 degrees.
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.h]: This documentation is for a getter, not a setter, so it should talk about what it means to see a longitude beyond ±180°:

> If a longitude is less than −180 degrees or greater than 180 degrees, the visible bounds straddles the antimeridian or international date line.

**Comment by fabian-guerra** [comment]
@1ec5 you are right. This PR does not fix #9731 or #4664 per se. Attempts only to clarify how to handle this use case.

**Pull request merged**